### DISCUSSION TITLE
How to query odb database

### DISCUSSION THREAD

**Discussion by titan73** [The-OpenROAD-Project/OpenROAD#5312] [Q&A] [answered]
I want to directly query odb database. How can I do that? There is no command for that.
I found some ord::xxx commands by searching the net but there is no proper documentation with all commands.

**Comment by titan73** [comment]
also found odb::xxx. No documentation either.
Most functions required a self argument.
Tested "::odb::dbBlock_getPowerDomains [::odb::dbBlock]" to get power domains but it crashes.

**Comment by titan73** [comment]
I got to something using orf and odb to get the created power domains:

foreach pd [::odb::dbBlock_getPowerDomains [::ord::get_db_block]] {
    puts "[::odb::dbPowerDomain_getName $pd]"
}

I think I get the idea. A pity it's not documented.

**Comment by maliberty** [comment, marked as answer]
We use swig on https://github.com/The-OpenROAD-Project/OpenROAD/blob/master/src/odb/include/odb/db.h so the APIs and comments there should be a guide.  If you have specific questions please ask.

**Comment by titan73** [comment]
Ok thanks. It helps. Although not all methods have comments.  dbPowerDomain for instance.
How does int& is supposed to be mapped in tcl? Ex: bool getArea(int& x1, int& y1, int& x2, int& y2);

**Reply by maliberty** [reply]
Some of these methods might need some swig configuration to make them more usable in TCL.   I can look at getArea

**Reply by maliberty** [reply]
I made some improvements for this api in https://github.com/The-OpenROAD-Project/OpenROAD/pull/5323
